components/org.wso2.carbon.identity.sso.saml/src/main/java/org/wso2/carbon/identity/sso/saml/servlet/SAMLSSOProviderServlet.java

```diff
@@ -366,9 +397,40 @@ private void handleIdPInitSSO(HttpServletRequest req, HttpServletResponse resp,
                     sendToFrameworkForLogout(req, resp, null, null, sessionId, true, isPost);
                 } else {
                     String errorResp = signInRespDTO.getResponse();
+                    String acsUrl = signInRespDTO.getAssertionConsumerURL();
+                    if (StringUtils.isBlank(acsUrl)) {
+                        String issuer = signInRespDTO.getIssuer();
+                        String returnToUrl = signInRespDTO.getReturnToURL();
+
+                        if (StringUtils.isBlank(issuer) && req.getParameter("spEntityID") != null) {
+                            issuer = req.getParameter("spEntityID");
+                        }
+                        if (StringUtils.isBlank(returnToUrl) && req.getParameter("returnTo") != null) {
+                            returnToUrl = req.getParameter("returnTo");
+                        }
+                        if (StringUtils.isNotBlank(issuer)) {
+                            SAMLSSOServiceProviderDO serviceProviderDO = SAMLSSOUtil.getSPConfig(SAMLSSOUtil
+                                            .getTenantDomainFromThreadLocal(),
+                                    SAMLSSOUtil.splitAppendedTenantDomain(issuer));
+                            //for IDP init SLO, priority should be given to log out URL.
+                            acsUrl = serviceProviderDO.getSloResponseURL();
+                            if (StringUtils.isBlank(acsUrl)) {
+                                acsUrl = serviceProviderDO.getDefaultAssertionConsumerUrl();
+                            }
+
+                            //check whether ReturnToUrl query param is included in the configured Urls.
+                            if (StringUtils.isNotBlank(returnToUrl)) {
+                                List<String> returnToUrls = serviceProviderDO.getIdpInitSLOReturnToURLList();
+                                if (returnToUrls.contains(returnToUrl)) {
```

Comment:
This check is quite rigid. Fro ACS urls we allow regex based expressions. Should we think about that in here as well?

Reply:
Regex based expressions for ACS are not supported in SAML.